docs/decisions/0031-agents.md

```diff
@@ -0,0 +1,277 @@
+
+[Playground](https://platform.openai.com/playground)
+
+## Unified Interface And Data Contract
+To enable collaboration and customization scenarios, further details of which will be provided below, agents should have a unified interface. They should inherit from the same abstract Agent class and have one data type - AgentMessage, for both input and output data contracts. This will enable a very powerful and extensible model, allowing the addition of any agent to any chat/collaboration, regardless of its implementation, and enabling a seamless chat conversation where the agent's output message is added to the chat without any conversion, and the chat is sent as input for the next agent without any conversion as well.
+
+For example, the scenario where two agents need to communicate together can configured like this:
+```C#
+class AgentTurnBasedChat(Agent[] agents)
+{
+    async Task<AgentMessage[]> StartConversationAsync(AgentMessage[] messages)
+    {
+        var chat = new List<AgentMessage>(messages);
+
+        var nextAgentIndex = 0;
+
+        while (chatExitCondition)
+        {
+            var nextAgent = agents[nextAgentIndex];
+
+            var result = await nextAgent.InvokeAsync(chat); // Agents accept a list of the 'AgentMessage' class, so chat can be passed without any conversion required.  
+   
+            chat.AddRange(result); // The agent's result can be added back to the chat as is, without any conversion needed.
+
+            nextAgentIndex = ...;
+        }
+
+        return chat.ToArray();
+    }
+}
+
+
+Agent copywriter = new ChatCompletionAgent(name: "Mike", instructions: "You're a helpful ....", ...); // Agent using the Chat Completion API
+
+Agent artDirector = new OpenAIAssistant(name: "Roy", instructions: ...); // Agent using OpenAI Assistance API.
+
+AgentTurnBasedChat chat = new(agents: new {copywriter, artDirector}); // Any agent can be added to the chat as long as it inherits from the 'Agent' class.
+
+var result = await chat.StartConversationAsync(new[] { new AgentMessage(role: "user", "collaborate on advertising campaigns for out latest product ....") });
+
+```
+Pros:  
+- No changes are needed for existing agent chat/collaboration classes to accommodate new agents.  
+- No custom logic is required in the chat/collaboration classes to support new agents.
+   
+Cons:  
+- Some features of certain agents might not be fully utilized, as chat/collaboration classes are only aware of the unified interface supported by all agents.
+
+## Agent Collaboration
+There are three identified methods for agents to collaborate with each other so far. As new collaboration strategies are discovered, it should be relatively easy to support them. One prerequisite for this is to ensure that agents remain unaware of the conversations they participate in, maintaining a one-way dependency - the chat/collaboration blocks know about the agents, but the agents do not know about them.
+
+### Collaboration Chat
```

Comment:
nit: Feel free to skip if the naming is not important here.
Taking into account, that the name of next approach is `Turn-Based Chat`, and this name briefly describes the approach, in case of `Collaboration Chat` name, it's not clear that it assumes the presence of admin. For clear terminology, I would probably call it something like `Moderated Chat`.

Reply:
The names are just for example purposes and might change the implementation phase.